Test execution changes for Rake, Travis, and Appveyor

miah opened the pull request:
I made default task now perform lint, and 'test'
'test' now does unit and functional. Sorry executing tests is going to be slow, but we won't miss anything.
You can still use `rake test:functional` and `rake test:unit`
Wrapped test coverage setup in an environment variable named CI_ENABLE_COVERAGE
You can now set 'W' and 'V' environment variables to enable Warnings, and Verbose output during testing.

I've also reconfigured travis. 
I removed a bunch of old deploy logic.
Sorted ruby versions newest to oldest.
Jobs in the matrix will automatically use the newest ruby after the sort.
Added a coverage job for coveralls reporting that is allowed to fail.
Added a macos builder.

Because we no longer have 3 functional test tasks for windows anymore the appveyor configuration is greatly simplified.

miah committed:
Rebase master

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Remove all old deploy stuff

Sort Ruby newest to oldest, this lets jobs in the matrix automatically
set the newest.
Add coverage job that is allowed to fail.
Add Mac and Windows testing. (curious)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Clean up Rakefile

Use GLOBS
Default tasks are lint and test
test task now does unit and functional

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Wrap coverage in environment variable CI_ENABLE_COVERAGE

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Remove windows, it doesnt support ruby yet :(

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Linux is default, but if its not listed in OS only osx will be tested.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Forgot to rename these =)

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Move tasks under test namespace, accept license

moved lint, and 'test' under the `test:` namespace.
moved 'test' to `test:default`, lint is now `test:lint`.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Must define the env at top level or it won't join matrix.

Also enable fast finish so we don't wait on allowed fails before
returning build status.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
unsure if this works..

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Exclude osx coverage build

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Remove a un-needed linux build, make coverage SLOW=1
Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
coverage should be slow=1 now!

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Maybe stop coverage from running early.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Fix indent

Signed-off-by: Miah Johnson <miah@chia-pet.org>

zenspider commented on test/helper.rb:
```diff
@@ -8,19 +8,21 @@
-  add_group 'Resources', 'lib/resources'
-  add_group 'Matchers', 'lib/matchers'
-  add_group 'Backends', 'lib/inspec/backend'
+  require 'simplecov'
```
Indentation starts weird? Did you mean to wrap this in the same CI ENV?

miah replied:
Yes its weird right now. I'm having a issue where coverage seems to be running early. If you look at the travis build you'll see it 'send to coveralls' just before it starts executing the tests. If you look at the coveralls link you can see that its only a few files from the repo.

miah replied:
Look here: https://travis-ci.org/inspec/inspec/jobs/534128587#L2977

miah replied:
Indentation is weird because I've moved the ENV conditional around it a few different ways.

zenspider commented on Rakefile:
```diff
@@ -121,31 +112,25 @@ namespace :test do
   end
   # Inject a prerequisite task
-  task :'functional:windows' => [:accept_license]
+  task :unit => [:accept_license]
```
I'd like to get rid of `:accept_license` at some point. The tests should do what they need to do themselves.

miah replied:
I agree, at the very least can we make it automatically part of all the tasks in a namespace?

zenspider commented on Rakefile:
```diff
@@ -121,31 +112,25 @@ namespace :test do
     ])
-    t.warning = false # This just complains about things in underlying libraries
+    t.warning = true
     t.verbose = true
```
I screwed up with the ENV["V"] above. It should be everywhere. Can you fix that?

zenspider commented on Rakefile:
```diff
@@ -121,31 +112,25 @@ namespace :test do
       'lib/plugins/inspec-*/test/functional/**/*_test.rb',
     ])
-    t.warning = false # This just complains about things in underlying libraries
+    t.warning = true
```
YESSSS. Thank you. That comment is so not true. While this is gonna be gross, it'll be worth it.

zenspider commented on appveyor.yml:
```diff
@@ -67,32 +61,3 @@ test_script:
-      - SET SPEC_OPTS=--format progress
-      - SET WORKER_NUMBER=3
-      - SET SLOW=1
-      - bundle exec rake test:functional:windows
```
This is also gonna be a slog... but also worth it in the long run. Thank you.

miah replied:
Ya, the times aren't great now but will improve =)

zenspider reviewed (changes requested):
One thing pointed out for fixing in test/helper.rb and one minor request for Rakefile

miah committed:
All rake task verbose toggle through ENV['V']

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
remove `test` from the rake command

Sort the ci coverage job, maybe we can match without the script: as
well?

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
travis lint cleanup

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Allow warning disable via environment

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
remove deprecated www release tasks

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Run all the tests.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Don't update rubygems system before installing rubygems

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Put pry with test tools, remove useless group

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
All the kitchen gems are in integration group, we need this..
Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Open up restrictions on these gems as they were.. out of date.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
berkshelf is making dependency fun. bye.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
test-kitchen currently requires berkshelf

I also discussed with Jerry if we can replace Berks with policyfile for
our testing and it seems unlikely at this point. Will re-evaluate in the
future.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Our site uses InSpec to generate some docs, ensure license is accepted

Signed-off-by: Miah Johnson <miah@chia-pet.org>

zenspider reviewed (approved):
Approved pending getting travis green. Right now it is just complaining about style issues.

Thank you!

miah committed:
We need berkshelf for test-kitchen, but we need chef 14 to support ruby
2.4

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
rubocop

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
After further discussion, lets allow Warning to be toggled on by ENV
until things are better.

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
Ensure Rubygems is at least version 3.0

Signed-off-by: Miah Johnson <miah@chia-pet.org>

miah committed:
doesn't like the ~>

Signed-off-by: Miah Johnson <miah@chia-pet.org>

clintoncwolfe reviewed (approved):
Thanks!
